cmd/kubeadm/app/cmd/util/join_test.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"encoding/base64"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+type kubeConfigSpec struct {
+	CertificateAuthorityData string
+	CertificateAuthorityFile string
+	Server                   string
+	ClusterName              string
+	ContextName              string
+	User                     string
+}
+
+func generateKubeConfig(spec kubeConfigSpec) *clientcmdapi.Config {
+	cluster := clientcmdapi.NewCluster()
+
+	if spec.CertificateAuthorityData != "" {
+		decodedData, _ := base64.StdEncoding.DecodeString(spec.CertificateAuthorityData)
+		cluster.CertificateAuthorityData = decodedData
+	} else if spec.CertificateAuthorityFile != "" {
+		cluster.CertificateAuthority = spec.CertificateAuthorityFile
+	}
+
+	cluster.Server = spec.Server
+
+	context := clientcmdapi.NewContext()
+	context.Cluster = spec.ClusterName
+	context.AuthInfo = spec.User
+
+	return &clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{
+			spec.ClusterName: cluster,
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			spec.ContextName: context,
+		},
+		CurrentContext: spec.ContextName,
+	}
+}
+
+func writeKubeConfigToFile(config *clientcmdapi.Config, filePath string) error {
+	configBytes, err := clientcmd.Write(*config)
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(filePath, configBytes, 0644)
+}
```

Comment:
just move this code in inside the test (i.e. unroll it) because it's used in only one place now.